internal/dot/graph.go

```diff
@@ -252,6 +289,116 @@ func (dg *Graph) addToGroup(r *Result, id CtorID) {
 	group.Results = append(group.Results, r)
 }
 
+// Prune removes elements from the graph that do not have failed results.
+// Removing elements that do not have failing results makes the graph easier to debug, since non-failing
+// nodes and edges can clutter the graph and don't help the user debug.
+func (dg *Graph) Prune() {
+	dg.indexParams()
+	dg.pruneCtors()
+	dg.pruneGroups()
+}
+
+// Remove constructors from the graph that do not have failing Results.
+func (dg *Graph) pruneCtors() {
+	var pruned []*Ctor
+	for _, c := range dg.Ctors {
+		if _, ok := dg.Failed.ctors[c.ID]; ok {
+			pruned = append(pruned, c)
+		} else {
+			dg.pruneCtorParams(c)
+			dg.pruneGroupResults(c)
+			delete(dg.ctorMap, c.ID)
+		}
+	}
+
+	dg.Ctors = pruned
+}
+
+// indexParams builds a map of params to constructors that consume those params so that constructor results
+// can be associated with constructors that consume those results.
+func (dg *Graph) indexParams() {
+	pIndex := make(map[nodeKey][]*Ctor)
```

Comment:
Don't need to build this at the end. You can build this as constructors are added. In fact, this is replicating what the Container does:

https://github.com/uber-go/dig/blob/007ab720a796b1027b6c289fb4e836c8fc077357/dig.go#L121

So you can definitely track this as you go.

Reply:
Makes sense! Done.